rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.create/20_synthetic_source.yml

```diff
@@ -1214,3 +1224,231 @@ fallback synthetic_source for text field:
+                type: keyword
+                copy_to: copy
+              copy:
+                type: keyword
+
+  - do:
+      index:
+        index: test
+        id: 1
+        refresh: true
+        body:
+          name: "A"
+          copy: "world"
+          k: "hello"
+
+  - do:
+      index:
+        index: test
+        id: 2
+        refresh: true
+        body:
+          name: "B"
+          k: ["5", "6"]
+          copy: ["7", "8"]
+
+  - do:
+      search:
+        index: test
+        sort: name
+        body:
+          docvalue_fields: [ "copy" ]
+
+  - match:
+      hits.hits.0._source:
+        name: "A"
+        k: "hello"
+        copy: "world"
+  - match: { hits.hits.0.fields.copy: ["hello", "world"] }
+
+  - match:
+      hits.hits.1._source:
+        name: "B"
+        k: ["5", "6"]
+        copy: ["7", "8"]
+  - match: { hits.hits.1.fields.copy: ["5", "6", "7", "8"] }
+
+---
+synthetic_source with copy_to field from dynamic template having values in source:
+  - requires:
+      cluster_features: ["mapper.source.synthetic_source_with_copy_to_and_doc_values_false"]
+      reason: requires copy_to support in synthetic source
+
+  - do:
+      indices.create:
+        index: test
+        body:
+          mappings:
+            _source:
+              mode: synthetic
+            dynamic_templates:
+              - copy_template:
+                  match: "k"
+                  mapping:
+                    type: keyword
+                    copy_to: copy
+            properties:
+              name:
+                type: keyword
+              copy:
+                type: keyword
+
+  - do:
+      index:
+        index: test
+        id: 1
+        refresh: true
+        body:
+          name: "A"
+          k: "hello"
+
+  - do:
+      index:
+        index: test
+        id: 2
+        refresh: true
+        body:
+          name: "B"
+          copy: "world"
+          k: "hello"
+
+  - do:
+      index:
+        index: test
+        id: 3
+        refresh: true
+        body:
+          name: "C"
+          k: ["5", "6"]
+
+  - do:
+      index:
+        index: test
+        id: 4
+        refresh: true
+        body:
+          name: "D"
+          k: ["5", "6"]
+          copy: ["7", "8"]
+
+  - do:
+      search:
+        index: test
+        sort: name
+        body:
+          docvalue_fields: [ "copy" ]
+
+  - match:
+      hits.hits.0._source:
+        name: "A"
+        k: "hello"
+  - match: { hits.hits.0.fields.copy: ["hello"] }
+
+  - match:
+      hits.hits.1._source:
+        name: "B"
+        k: "hello"
+        copy: "world"
+  - match: { hits.hits.1.fields.copy: ["hello", "world"] }
+
+  - match:
+      hits.hits.2._source:
+        name: "C"
+        k: ["5", "6"]
+  - match: { hits.hits.2.fields.copy: ["5", "6"] }
+
+  - match:
+      hits.hits.3._source:
+        name: "D"
+        k: ["5", "6"]
+        copy: ["7", "8"]
+  - match: { hits.hits.3.fields.copy: ["5", "6", "7", "8"] }
+
```

Comment:
Maybe add a test where copy_to is applied in a context where the source has been stored already, e.g. in an array that has been stored.